### PR TITLE
NGSTACK-673: DX: Generate URLs directly

### LINKED ISSUE #91: Add `url_alias` property to `Location`

### PR DESCRIPTION
**Pull request by pspanja**
Closes https://github.com/netgen/ezplatform-site-api/issues/91

See https://github.com/netgen/ezplatform-site-api/issues/91 for the original issue and discussion.

A new `Url` object is added to the Site API `Content` and `Location` objects, making it possible to generate an URL directly, in PHP:

```php
$urlString = (string) $location->url;
$urlString = $location->url->get();
$urlString = $location->url->get(['siteaccess' => 'adminui', 'page' => 1, '_fragment' => 'frag');
$pathString = (string) $location->path;
$pathString = $location->path->getAbsolute();
$pathString = $content->path->getNetwork();
$pathString = $content->path->getRelative();
```

And in Twig:

```twig
{{ location.url }}
{{ location.url({siteaccess: 'adminui', page: 1, _fragment: 'frag'}) }}
{{ location.path }}
{{ location.path.absolute }}
{{ content.path.network }}
{{ content.path.relative }}
```
URL types directly correspond to what is defined in [Symfony's UrlGeneratorInterface](https://github.com/symfony/routing/blob/6.2/Generator/UrlGeneratorInterface.php) and RFC 3986.

Both `Url` and `Path` objects implement `__toString`, which will generate absolute URL and absolute URL path for the current siteaccess, respectively.

`Content` and `Location` objects also implement public methods `getUrl(array $parameters = []): string` and `getPath(array $parameters = []): string`, which return absolute URL and absolute URL path respectively, and make possible to write in PHP:

```php
$urlString = $location->getUrl(['siteaccess' => 'adminui', 'page' = 1, '_fragment': 'frag');
$urlPath = $location->getPath(['siteaccess' => 'adminui', 'page' = 1, '_fragment': 'frag');
```

And Twig:

```twig
{{ location.url({siteaccess: 'adminui', page: 1, _fragment: 'frag'}) }}
{{ location.path({siteaccess: 'adminui', page: 1, _fragment: 'frag'}) }}
```

## To do
- [x] Fix tests
- [x] Write documentation

### CODE REVIEW

File: bundle/Resources/views/pagelayout.html.twig [hunk 1]
```diff
@@ -1,11 +1,14 @@
...
 <!doctype html>
 <html lang="{{ app.request.locale|replace({'_': '-'}) }}">
 <head>
     <meta charset="utf-8">
-    <title>{{ content.name|default( 'Home' ) }}</title>
+    <title>{{ content.name }}</title>
     <meta name="generator" content="Netgen Site API"/>
-    {% if content is defined and content.mainLocation is not null %}
-        <link rel="canonical" href="{{ ibexa_path( content.mainLocation ) }}" />
+    {% if location.url is not null %}
```
Comment (emodric) [inline bundle/Resources/views/pagelayout.html.twig]: Pagelayout is used when there's no content and location defined, in custom modules and such. So the `is defined` check would still be needed.

File: bundle/Resources/views/default/content/text_linked.html.twig [hunk 2]
```diff
@@ -2,7 +2,7 @@
 {# location \Netgen\IbexaSiteApi\API\Values\Location #}
 
 {% if location is defined %}
-    <a href="{{ ibexa_path(location) }}">{{ content.name }}</a>
+    <a href="{{ location.url }}">{{ content.name }}</a>
```
Comment (emodric) [inline bundle/Resources/views/default/content/text_linked.html.twig]: This is quite a change in nomenclature here. We all know what `path` and `url` are in Symfony, and changing this here could cause confusion. I would rather have two properties `location.path` and `location.url` and then further refine `location.path` with `location.path.relative`, `location.path.absolute` and `location.path.network_relative`.

Comment (iherak) [reply bundle/Resources/views/default/content/text_linked.html.twig]: I second this, it would be better to be in sync with Symfony nomenclature.

Comment (emodric) [reply bundle/Resources/views/default/content/text_linked.html.twig]: Now I realize that `location.path` is taken by Ibexa location already, but I would be fine overriding it in Site API. If you need access to it, you would use `location.innerLocation.path` or even using a renamed property `location.pathArray`

Comment (pspanja) [reply bundle/Resources/views/default/content/text_linked.html.twig]: Good points, will update 👍

File: bundle/Resources/views/pagelayout.html.twig (hunk 1 above)
Comment (pspanja) [reply bundle/Resources/views/pagelayout.html.twig]: Fixed in https://github.com/netgen/ibexa-site-api/pull/11/commits/59eedfdf273936da2cdc10397bc92ca1c13ae898.

File: bundle/Resources/views/default/content/text_linked.html.twig (hunk 2 above)
Comment (pspanja) [reply bundle/Resources/views/default/content/text_linked.html.twig]: Changed in https://github.com/netgen/ibexa-site-api/pull/11/commits/59eedfdf273936da2cdc10397bc92ca1c13ae898 and updated the description, is this what you had in mind?

Comment (emodric) [reply bundle/Resources/views/default/content/text_linked.html.twig]: Yep, looks good! 👍

Comment (pspanja) [reply bundle/Resources/views/default/content/text_linked.html.twig]: Great, then the PR on the `netgen/site-bundle` for `Location::$path` => `Location::$pathArray` change follows.

File: bundle/Resources/views/default/content/embed.html.twig [hunk 3]
```diff
@@ -2,7 +2,7 @@
 {# location \Netgen\IbexaSiteApi\API\Values\Location #}
 
 {% if location is defined %}
-    <h3><a href="{{ ibexa_path(location) }}">{{ content.name }}</a></h3>
+    <h3><a href="{{ location.url }}">{{ content.name }}</a></h3>
```
Comment (emodric) [inline bundle/Resources/views/default/content/embed.html.twig]: `location.path` ?

File: bundle/Resources/views/default/content/embed.html.twig [hunk 4]
```diff
@@ -2,7 +2,7 @@
...
 {% else %}
-    <h3><a href="{{ ibexa_path(content) }}">{{ content.name }}</a></h3>
+    <h3><a href="{{ content.url }}">{{ content.name }}</a></h3>
```
Comment (emodric) [inline bundle/Resources/views/default/content/embed.html.twig]: Here too.

File: bundle/Resources/views/default/content/text_linked.html.twig [hunk 5]
```diff
@@ -2,7 +2,7 @@
...
 {% else %}
-    <a href="{{ ibexa_path(content) }}">{{ content.name }}</a>
+    <a href="{{ content.url }}">{{ content.name }}</a>
```
Comment (emodric) [inline bundle/Resources/views/default/content/text_linked.html.twig]: Here too, in two locations should be `content.path` and `location.path`

File: docs/reference/objects.rst [hunk 6]
```diff
@@ -434,6 +434,72 @@ Used to filter field relation Locations from the `Field`_ with the given ``$iden
...
+|                                        |         '_fragment' => 'frag'                                                      |
+|                                        |     ]);                                                                            |
+|                                        |                                                                                    |
++----------------------------------------+------------------------------------------------------------------------------------+
+| **Example in Twig**                    | .. code-block:: twig                                                               |
+|                                        |                                                                                    |
+|                                        |     {% set absolute_path = content.path() %}                                       |
+|                                        |     {% set absolute_path = content.path({                                          |
+|                                        |         siteaccess: 'adminui',                                                     |
```
Comment (emodric) [inline docs/reference/objects.rst]: Maybe use `admin` in docs, since `adminui` is specific to Media Site?

File: lib/API/Values/Location.php [hunk 7]
```diff
@@ -24,11 +24,13 @@
  * @property-read string $remoteId
  * @property-read int $parentLocationId
  * @property-read string $pathString
- * @property-read int[] $path
+ * @property-read int[] $pathArray
```
Comment (emodric) [inline lib/API/Values/Location.php:27]: Since this is a minor breaking change, should we consider bumping the major version?

Comment (pspanja) [reply lib/API/Values/Location.php:27]: Yes indeed, this will be released as version 6.

File: docs/reference/objects.rst (hunk 6 above)
Comment (pspanja) [reply docs/reference/objects.rst]: Updated in https://github.com/netgen/ibexa-site-api/pull/11/commits/f0358646d4effec9171ddfca544c6d6189c0b537

File: bundle/Resources/views/default/content/text_linked.html.twig (hunk 5 above)
Comment (pspanja) [reply bundle/Resources/views/default/content/text_linked.html.twig]: Updated in https://github.com/netgen/ibexa-site-api/pull/11/commits/c734f615167a5ab8333947c44ff8c2fd41448bdb

File: bundle/Resources/views/default/content/embed.html.twig (hunk 4 above)
Comment (pspanja) [reply bundle/Resources/views/default/content/embed.html.twig]: Updated in https://github.com/netgen/ibexa-site-api/pull/11/commits/c734f615167a5ab8333947c44ff8c2fd41448bdb

File: bundle/Resources/views/default/content/embed.html.twig (hunk 3 above)
Comment (pspanja) [reply bundle/Resources/views/default/content/embed.html.twig]: Updated in https://github.com/netgen/ibexa-site-api/pull/11/commits/c734f615167a5ab8333947c44ff8c2fd41448bdb

File: bundle/Resources/views/default/content/text_linked.html.twig (hunk 2 above)
Comment (pspanja) [reply bundle/Resources/views/default/content/text_linked.html.twig]: Created PRs in site-bundle and media-site:
- https://github.com/netgen/media-site/pull/172
- https://github.com/netgen/site-bundle/pull/49

**Review by iherak — APPROVED** [review]

**Comment by pspanja** [comment]
Thanks @emodric and @iherak!

**Pull request merged**